### PR TITLE
[CPU][INT4] Add AWQ frontend support for CPU 

### PR DESCRIPTION
**Pull request by jianan-gu**
This PR implements AWQ int4 frontends support for CPU, and is calling the CPU sgl-kernels in https://github.com/sgl-project/sglang/pull/8226

### CODE REVIEW

File: python/sglang/srt/layers/amx_utils.py [hunk 1]
```diff
@@ -6,6 +6,10 @@
 
 logger = logging.getLogger(__name__)
 
+import os
+
+SGLANG_USE_CPU_INT4_W4A8 = os.getenv("SGLANG_USE_CPU_INT4_W4A8", "0") == "1"
```
Comment (Alcanderian) [inline python/sglang/srt/layers/amx_utils.py]: get_bool_env_var

File: python/sglang/srt/layers/quantization/__init__.py [hunk 2]
```diff
@@ -88,13 +88,13 @@ def override_quantization_method(self, *args, **kwargs):
...
 VLLM_QUANTIZATION_METHODS = {
     "aqlm": AQLMConfig,
-    "awq": AWQConfig,
     "deepspeedfp": DeepSpeedFPConfig,
     "tpu_int8": Int8TpuConfig,
     "fbgemm_fp8": FBGEMMFp8Config,
     "marlin": MarlinConfig,
     "gguf": GGUFConfig,
     "gptq_marlin_24": GPTQMarlin24Config,
+    "awq": AWQConfig,
```
Comment (Alcanderian) [inline python/sglang/srt/layers/quantization/__init__.py:95]: Why the order changed?

File: python/sglang/srt/layers/quantization/awq.py [hunk 3]
```diff
@@ -109,7 +118,7 @@ def get_name(self) -> str:
         return "awq"
 
     def get_supported_act_dtypes(self) -> List[torch.dtype]:
-        return [torch.half]
+        return [torch.half, torch.bfloat16]
```
Comment (Alcanderian) [inline python/sglang/srt/layers/quantization/awq.py]: Is bf16 only available on cpu?

File: python/sglang/srt/layers/amx_utils.py (hunk 1 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: Thanks for this, have refined the codes with `get_bool_env_var`

File: python/sglang/srt/layers/quantization/__init__.py (hunk 2 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/quantization/__init__.py:95]: was following the order of awq/awq_marlin by mistake, have revert this order change

File: python/sglang/srt/layers/quantization/awq.py (hunk 3 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/quantization/awq.py]: yes, have refined here to add bf16 only on CPU

File: python/sglang/srt/configs/update_config.py [hunk 4]
```diff
@@ -23,19 +23,23 @@ def may_get_weight_block_size(model_config, load_config):
...
     if weight_block_size is not None:
         # See NOTE(HandH1998): To ensure proper alignment of the block-wise quantization scales, the output_size of the weights for both the gate and up layers must be divisible by block_n.
         assert (
-            len(weight_block_size) == 2
-        ), "Only len(weight_block_size) == 2 is supported"
-        assert (
-            weight_block_size[0] == weight_block_size[1]
-        ), "Only weight_block_size[0] == weight_block_size[1] is supported"
-
+            len(weight_block_size) == 2 or len(weight_block_size) == 1
```
Comment (mingfeima) [inline python/sglang/srt/configs/update_config.py]:
```suggestion
            len(weight_block_size) in [1, 2]
```

File: python/sglang/srt/layers/amx_utils.py [hunk 5]
```diff
@@ -2,10 +2,13 @@
...
 import torch
 
-from sglang.srt.utils import cpu_has_amx_support
+from sglang.srt.utils import cpu_has_amx_support, get_bool_env_var
 
 logger = logging.getLogger(__name__)
 
+import os
+
+SGLANG_USE_CPU_INT4_W4A8 = get_bool_env_var("SGLANG_USE_CPU_INT4_W4A8")
```
Comment (mingfeima) [inline python/sglang/srt/layers/amx_utils.py]: why we need an env to tell us to kick off int4 w4a8?

Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: Was considering providing a choice to allow a user configurable quantization-recipe-like flag to choose QGEMM type (default w4a16, w4a8 as option) to make tradeoff between runtime accuracy and performance. Not sure if any better approach here.

File: python/sglang/srt/configs/update_config.py [hunk 6]
```diff
@@ -23,19 +23,23 @@ def may_get_weight_block_size(model_config, load_config):
...
     if weight_block_size is not None:
         # See NOTE(HandH1998): To ensure proper alignment of the block-wise quantization scales, the output_size of the weights for both the gate and up layers must be divisible by block_n.
         assert (
-            len(weight_block_size) == 2
-        ), "Only len(weight_block_size) == 2 is supported"
-        assert (
-            weight_block_size[0] == weight_block_size[1]
-        ), "Only weight_block_size[0] == weight_block_size[1] is supported"
-
+            len(weight_block_size) in [1, 2]
```
Comment (mingfeima) [inline python/sglang/srt/configs/update_config.py]: when do we have weight_block_siz on only 1 dimension.

File: python/sglang/srt/layers/amx_utils.py (hunk 5 above)
Comment (mingfeima) [reply python/sglang/srt/layers/amx_utils.py]: OK let me think ...

By the way please which of the following did you implement？
- [ ] w4a8 GEMM
- [ ] w4a8 MoE
- [ ] w4a16 GEMM
- [ ] w4a16 MoE

File: python/sglang/srt/layers/amx_utils.py [hunk 7]
```diff
@@ -28,6 +31,106 @@ def dim_is_supported(weight):
     return OC % TILE_N == 0 and IC % TILE_K == 0
 
 
+def _autoawq_to_int4pack(
```
Comment (mingfeima) [inline python/sglang/srt/layers/amx_utils.py]: put `_autoawq_to_int4pack` in C++. You can leave it un-optimized (scalar code is fine), but it needs to be done in C++.

Since I plan to optimize weight loading/prepacking later on.

File: python/sglang/srt/layers/amx_utils.py [hunk 8]
```diff
@@ -28,6 +31,106 @@ def dim_is_supported(weight):
...
+        qweight = qweight.transpose(-3, -2)
+        # bit packing
+        BIT_COUNT = 32
+        qweight = qweight.reshape(-1, BIT_COUNT * 2)
+        qweight = (qweight[:, BIT_COUNT:] << 4) | qweight[:, :BIT_COUNT]
+        qweight = qweight.reshape(*dims, N, K // 2)
+        return qweight, qzeros, scales
+
+
+def _amx_process_int4_packed_qweight_after_loading(
```
Comment (mingfeima) [inline python/sglang/srt/layers/amx_utils.py]: can we fuse `_amx_process_int4_packed_qweight_after_loading` with `_amx_process_weight_after_loading` and call the later from awq.py?

File: python/sglang/srt/layers/amx_utils.py [hunk 9]
```diff
@@ -28,6 +31,106 @@ def dim_is_supported(weight):
...
+        prefix_list = weight_names[0].split("_")
+        # MoE layers have prefix
+        has_prefix = len(prefix_list) != 1
+        use_w4a8 = SGLANG_USE_CPU_INT4_W4A8 and not has_prefix
+        if use_w4a8:
+            # TODO: support MoE layers for W4A8 path
+            qweight, qzeros, scales, compensation = _autoawq_to_int4pack(
+                qweight_tensor.data, qzeros_tensor.data, scales_tensor.data, use_w4a8
+            )
+            compensation = torch.nn.Parameter(compensation, requires_grad=False)
```
Comment (mingfeima) [inline python/sglang/srt/layers/amx_utils.py]: compensation type of logic needs to be in C++, anyway put packing related logic in C++. The python API should be as simple as possible.

File: python/sglang/srt/layers/amx_utils.py [hunk 10]
```diff
@@ -28,6 +31,106 @@ def dim_is_supported(weight):
...
 
+def _autoawq_to_int4pack(
+    qweight: torch.Tensor, qzeros: torch.Tensor, scales: torch.Tensor, is_w4a8: bool
+):
+    """Convert AutoAWQ weight format to sgl-kernel's CPU int4
+
+    Args:
+        qweight: (*, K, N / 8), int32
+        qzeros: (*, K / group_size, N / 8), int32
+        scales: (*, K / group_size, N), bfloat16
```
Comment (mingfeima) [inline python/sglang/srt/layers/amx_utils.py]: you can :

* overload `convert_weight_packed`
* new api maybe `convert_weight_packed_scale_zp`
* simply skip packing scale and zp, use prefetch can also achieve good  performance and much simpler.

File: python/sglang/srt/layers/quantization/__init__.py [hunk 11]
```diff
@@ -112,7 +112,15 @@ def get_quantization_config(quantization: str) -> Type[QuantizationConfig]:
             f"Invalid quantization method: {quantization}. "
             f"Available methods: {list(QUANTIZATION_METHODS.keys())}"
         )
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/__init__.py:106]:
```suggestion
        )
        if is_cpu():
            return QUANTIZATION_METHODS[quantization]
```

we should also have a list e.g. `CPU_QUANTIZATIPON_METHODS` something like this, and do checks here.

File: python/sglang/srt/layers/quantization/__init__.py [hunk 12]
```diff
@@ -112,7 +112,15 @@ def get_quantization_config(quantization: str) -> Type[QuantizationConfig]:
...
-    if quantization in VLLM_QUANTIZATION_METHODS and not VLLM_AVAILABLE:
+
+    from sglang.srt.utils import is_cpu
+
+    # for CPU quantization, will not need vLLM dependency
+    if (
+        not is_cpu()
+        and quantization in VLLM_QUANTIZATION_METHODS
+        and not VLLM_AVAILABLE
+    ):
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/__init__.py]:
```suggestion
```

File: python/sglang/srt/layers/quantization/awq.py [hunk 13]
```diff
@@ -58,8 +62,13 @@
     )
 
     warnings.warn(f"HIP does not support fused_marlin_moe currently.")
+elif _is_cpu and _is_cpu_amx_available:
+    from sglang.srt.layers.amx_utils import (
+        SGLANG_USE_CPU_INT4_W4A8,
+        _amx_process_int4_packed_qweight_after_loading,
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/awq.py]: use `_amx_process_weight_after_loading` the same API for the same purpose.

File: python/sglang/srt/layers/quantization/awq.py [hunk 14]
```diff
@@ -394,16 +409,40 @@ def create_weights(
...
+                    layer.scales,
+                    layer.qzeros,
+                    layer.compensation,
+                    bias,
+                    x.dtype,
+                )
+            else:
+                return torch.ops.sgl_kernel.int4_scaled_mm_cpu(
+                    x, layer.qweight, layer.qzeros, layer.scales, bias
+                )
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/awq.py]: use the same kernel and pack compensation with qweight, just like int8 path.

File: python/sglang/srt/layers/quantization/awq.py [hunk 15]
```diff
@@ -669,9 +708,21 @@ def create_weights(
...
+        if _is_cpu:
+            assert (
+                _is_cpu_amx_available
+            ), "AWQLinearMethod on CPU requires that CPU has AMX support"
+            _amx_process_int4_packed_qweight_after_loading(
+                layer, ["w13_qweight", "w13_qzeros", "w13_scales"], "awq"
+            )
+            _amx_process_int4_packed_qweight_after_loading(
+                layer, ["w2_qweight", "w2_qzeros", "w2_scales"], "awq"
+            )
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/awq.py]: use `_amx_process_packed_qweight_after_loading`

File: python/sglang/srt/layers/quantization/awq.py [hunk 16]
```diff
@@ -748,6 +799,28 @@ def apply(
...
+            return torch.ops.sgl_kernel.fused_experts_cpu(
+                x,
+                layer.w13_qweight,
+                layer.w2_qweight,
+                topk_weights,
+                topk_ids,
+                False,  # inplace See [Note] inplace should be False in fused_experts.
+                False,  # use_int8_w8a8
+                False,  # use_fp8_w8a16
+                True,  # use_int4_w4a16
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/awq.py]: can we simplify these parameters, in future we may have more...

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/amx_utils.py (hunk 5 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: This PR (and kernels in https://github.com/sgl-project/sglang/pull/8226) contains w4a8 GEMM and  w4a16 GEMM/MoE

File: python/sglang/srt/configs/update_config.py (hunk 6 above)
Comment (jianan-gu) [reply python/sglang/srt/configs/update_config.py]: The Group size of AWQ (or GPTQ), only covers 1 dim (K). 
https://huggingface.co/Qwen/QwQ-32B-AWQ/blob/main/config.json#L20

File: python/sglang/srt/layers/amx_utils.py (hunk 7 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: Sure, have refine the frontend here, and will call C++ functions `torch.ops.sgl_kernel.convert_weight_packed_scale_zp `from https://github.com/sgl-project/sglang/pull/8226

File: python/sglang/srt/layers/amx_utils.py (hunk 8 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: Sure， have fused them together

File: python/sglang/srt/layers/amx_utils.py (hunk 9 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: Sure, have refine the frontend here, and will make the changes in https://github.com/sgl-project/sglang/pull/8226

File: python/sglang/srt/layers/amx_utils.py (hunk 10 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: Sure, will refine the pack function with `convert_weight_packed_scale_zp`

File: python/sglang/srt/layers/quantization/__init__.py (hunk 11 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/quantization/__init__.py:106]: yes, created `CPU_QUANTIZATIPON_METHODS `.

File: python/sglang/srt/layers/quantization/awq.py (hunk 13 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/quantization/awq.py]: sure, have refined here with `_amx_process_weight_after_loading`

File: python/sglang/srt/layers/quantization/awq.py (hunk 14 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/quantization/awq.py]: Yes, have refine the frontend here, and will make the kernel changes in https://github.com/sgl-project/sglang/pull/8226

File: python/sglang/srt/layers/quantization/awq.py (hunk 16 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/quantization/awq.py]: Sure, created `CPUMoECompMethod`， to work as a switch to choose different MoE computation mothod.

File: python/sglang/srt/layers/amx_utils.py (hunk 5 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/amx_utils.py]: updated w4a8 MoE, now we have all the four kernels.

File: python/sglang/srt/configs/update_config.py [hunk 17]
```diff
@@ -23,19 +23,24 @@ def may_get_weight_block_size(model_config, load_config):
...
-        assert (
-            len(weight_block_size) == 2
-        ), "Only len(weight_block_size) == 2 is supported"
-        assert (
-            weight_block_size[0] == weight_block_size[1]
-        ), "Only weight_block_size[0] == weight_block_size[1] is supported"
-
+        assert len(weight_block_size) in [
+            1,
+            2,
```
Comment (mingfeima) [inline python/sglang/srt/configs/update_config.py:32]: the format is a little odd here.

File: python/sglang/srt/layers/amx_utils.py [hunk 18]
```diff
@@ -6,6 +6,15 @@
...
 logger = logging.getLogger(__name__)
 
+from enum import IntEnum
+
+
+class CPUMoECompMethod(IntEnum):
+    BF16_GEMM = 0
+    INT8_W8A8_GEMM = 1
+    FP8_W8A16_GEMM = 2
+    INT4_W4A8_GEMM = 3
```
Comment (mingfeima) [inline python/sglang/srt/layers/amx_utils.py]:
```suggestion
class CPUQuantMethod(IntEnum):
    UNQUANT = 0
    INT8_W8A8 = 1
    FP8_W8A16 = 2
    INT4_W4A8 = 3
```

update C++ accordingly.

File: python/sglang/srt/layers/quantization/awq.py [hunk 19]
```diff
@@ -840,6 +881,29 @@ def apply(
...
+            topk_output = dispatch_output.topk_output
+            topk_weights, topk_ids, _ = topk_output
+            return torch.ops.sgl_kernel.fused_experts_cpu(
+                x,
+                layer.w13_qweight,
+                layer.w2_qweight,
+                topk_weights,
+                topk_ids,
+                False,  # inplace See [Note] inplace should be False in fused_experts.
+                CPUMoECompMethod.INT4_W4A8_GEMM,
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/awq.py]: change the naming, as we will apply this to both gemm and moe modules. and bfloat16 means unquant, aligns with sglang quantization recipe definition.

**Review by mingfeima — APPROVED** [review]
just some minor issues to address.

File: python/sglang/srt/layers/quantization/awq.py (hunk 19 above)
Comment (jianan-gu) [reply python/sglang/srt/layers/quantization/awq.py]: Got it, have refine the naming

File: python/sglang/srt/configs/update_config.py (hunk 17 above)
Comment (jianan-gu) [reply python/sglang/srt/configs/update_config.py:32]: format by pre-commit formats, while check after rebase

**Comment by jianan-gu** [comment]
The changes in this PR is merging into https://github.com/sgl-project/sglang/pull/22685, closing this PR.
